File: python/Q45_jump_game_ii.py
#
# @lc app=leetcode id=45 lang=python3
#
# [45] Jump Game II
#

# @lc code=start
class Solution:
    def jump(self, nums: List[int]) -> int:
        left: int = 0
        right: int = 0
        result: int = 0

        while right < len(nums) - 1:
            max_length: int = 0
            for i in range(left, right + 1):
                if (new_length := i + nums[i]) > max_length:
                    max_length = new_length

            left = right + 1
            right = max_length
            result += 1

        return result


# @lc code=end

File: rust/src/leetcode/Q45_jump_game_ii.rs
/*
 * @lc app=leetcode id=45 lang=rust
 *
 * [45] Jump Game II
 */

// @lc code=start
impl Solution {
    pub fn jump(nums: Vec<i32>) -> i32 {
        let mut result: i32 = 0;
        let mut left: usize = 0;
        let mut right: usize = 0;

        while right < nums.len() - 1 {
            let mut right_most: usize = left;

            for i in left..=right {
                right_most = right_most.max(nums[i] as usize + i);
            }

            result += 1;
            left = right + 1;
            right = right_most;
        }

        result
    }
}
// @lc code=end
